lib/SILOptimizer/Utils/Differentiation/ADContext.cpp

```diff
@@ -54,9 +54,17 @@ static FuncDecl *findOperatorDeclInProtocol(DeclName operatorName,
 // ADContext methods
 //===----------------------------------------------------------------------===//
 
+static SynthesizedFileUnit *createNewSynthesizedFileUnit(ModuleDecl *module) {
+  auto &ctx = module->getASTContext();
+  auto *file = new (ctx) SynthesizedFileUnit(*module);
+  module->addFile(*file);
+  return file;
+}
+
 ADContext::ADContext(SILModuleTransform &transform)
     : transform(transform), module(*transform.getModule()),
-      passManager(*transform.getPassManager()) {}
+      passManager(*transform.getPassManager()),
+      synthesizedFile(*createNewSynthesizedFileUnit(module.getSwiftModule())) {}
```

Comment:
Currently, `ADContext` in the differentiation transform is responsible for creating a `SynthesizedFileUnit`.

Instead, we could add utilities for creating a `SynthesizedFileUnit` on `ModuleDecl`:
* `Optional<SynthesizedFileUnit *> ModuleDecl::getSynthesizedFileUnit`
* `SynthesizedFileUnit *ModuleDecl::getOrCreateSynthesizedFileUnit`

`SynthesizedFileUnit` is a new construct, so we could establish invariants like "each module only has one synthesized file unit". I don't know about benefits of these invariants if `SynthesizedFileUnit` is used only by the differentiation transform.